### PR TITLE
IA-3120 fix createAzureVm undo step

### PR DESCRIPTION
**Pull request by Qi77Qi**
https://broadworkbench.atlassian.net/browse/IA-3120

When we delete VM, we also need to delete OS disk and nic automatically

### CODE REVIEW

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java [hunk 1]
```diff
@@ -177,9 +177,22 @@ public StepResult undoStep(FlightContext context) throws InterruptedException {
     ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
 
     try {
+      VirtualMachine resolvedVm =
+              computeManager
+                      .virtualMachines()
+                      .getByResourceGroup(
+                              azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+
       computeManager
```
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java:186]: Is this delete idempotent? We don't want any exceptions coming from undo if it's already deleted.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java [hunk 2]
```diff
@@ -177,9 +177,22 @@ public StepResult undoStep(FlightContext context) throws InterruptedException {
...
-          .deleteByResourceGroup(azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+              .virtualMachines()
+              .deleteByResourceGroup(azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+
+      resolvedVm
+              .networkInterfaceIds()
+              .forEach(nic -> computeManager.networkManager().networkInterfaces().deleteById(nic));
+
+      // Delete the OS disk
+      computeManager.disks().deleteById(resolvedVm.osDiskId());
```
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java:195]: I wonder if you'll need to keep coming back here to add things to be deleted. That would be easy to miss, so I wonder if you wouldn't be better served by a central `deleteVmAndEverythingInIt()` method, especially if you need to do this in another place, like a delete endpoint.

**Review by jaycarlton — APPROVED** [review]

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java [hunk 3]
```diff
@@ -177,9 +177,22 @@ public StepResult undoStep(FlightContext context) throws InterruptedException {
...
+                      .virtualMachines()
+                      .getByResourceGroup(
+                              azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+
       computeManager
-          .virtualMachines()
-          .deleteByResourceGroup(azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+              .virtualMachines()
+              .deleteByResourceGroup(azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+
```
Comment (ddietterich) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java:189]: I am not sure I understand the dependencies, but I think this undo code is not idempotent. If the pod crashed (or was stopped by K8s) at this point, then a rerun of this undo would not be able to locate the nics or the disks. The VM would be deleted so the retrieval at line 180 would not find it.

The fix would be to find the objects by name, the same way you find them in the `do()` code.

**Review by ddietterich — CHANGES_REQUESTED** [review]

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java (hunk 2 above)
Comment (Qi77Qi) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java:195]: where would you but this central method

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java (hunk 3 above)
Comment (Qi77Qi) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java:189]: good point..I can fix it for nic, but OS disk doesn't have a known name, so Idk if we can fix this completely

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java (hunk 2 above)
Comment (jaycarlton) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java:195]: Wherever your delete endpoint is, or a utility class. If you only ever need it here, then there's no need to break it out.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java (hunk 3 above)
Comment (ddietterich) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java]: Probably my misunderstanding about the disk. I see at line 85, a disk is retrieved by name. Is that separate from the OS disk? Or is that the disk you are supposed to delete?

If the OS disk is automatically created by Azure, but not automatically deleted by Azure, I don't see a way to fix the idempotency.

Comment (Qi77Qi) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java]: >Probably my misunderstanding about the disk. I see at line 85, a disk is retrieved by name. Is that separate from the OS disk? Or is that the disk you are supposed to delete?

yea OS disk and the user disk are 2 disks. We define user disk but not OS disk.

>If the OS disk is automatically created by Azure, but not automatically deleted by Azure, I don't see a way to fix the idempotency.
yes.

Comment (ddietterich) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java]: OK, then would you document the problem case (orphan NIC and disk in a small failure window) in the undo code for our future selves? I'll approve the PR now.

**Review by ddietterich — APPROVED** [review]

**Comment by Qi77Qi** [comment]
@ddietterich  nic should be good now...OS disk is the only thing, I have a TODO for now...will look at this a bit more once I get the main ticket I'm working on in a good state

**Pull request merged**